### ISSUE TITLE
Thanos compcator & receiver :: vertical compaction not working as expected

### ISSUE THREAD

**Issue by momoadc** [thanos-io/thanos#6004]
Thanos version: 0.29.0
Prometheus version: 2.36.2
OpenShift version: 4.11.13
**Thanos, Prometheus and Golang version used**:

**Object Storage Provider**: Dell Powerscale S3 implementation

**What happened**: I'm using a setup of federated Prometheis that scrapes data from the OpenShift monitoring stack (HA pair of Prometheis) and remote writes it to a setup of receivers (routers load-balancing to ingestors by hashring).
I'm running a compactor with vertical compaction enabled on prometheus_replica labels (each set of federated Prometheis have a different replica label name, prometheus_replica, prometheus_hub_replica and prometheus_tenant_replica). The streams in the bucket are divided only by receiver replica and tenant (the same tenant is used in each stream) so that metrics federated from OpenShift's Prometheis are written to the receiver via the same Prometheus and end up in the same stream.
The problem I'm issuing is that vertical compaction (using the penalty function) not working as expected and only compact blocks from different receivers.

**What you expected to happen**: I'm expecting that vertical compaction will deduplicate metrics from OpenShift Prometheis HA pair but it is not working

**How to reproduce it (as minimally and precisely as possible)**: create a setup of federated Prometheis so that the Prometheus in charge of remote writing the data to the receiver aggregates metrics from HA pair of Prometheis. Activate vertical compaction with penalty function with deduplication-label=prometheus_replica

**Anything else we need to know**: I'm working in a restricted network (DMZ) and cannot fetch logs/screenshots. I'll happily supply additional information about the stack if needed. 
I would like to understand if vertical compaction can happen within the same stream. To my understanding, The receiver will upload the blocks to S3 with external labels defined statically or via tenant's HTTP header (is it possible to upload additional labels based on external labels in the metrics)

**Comment by yeya24** [comment]
Please share the configuration of your thanos compactor, receivers and also external labels configured on your prometheuses.

**Comment by v1jayr** [comment]
I'm having similar issue with my Thanos setup. We are thinking about having HA prometheus remote write to Thanos Receiver.  Based on the docs, the series from the Prometheus could be spread across multiple receiver nodes producing their own TSDB blocks. Vertical compaction run by Compactor does identifies the blocks from each receiver is different stream and hence compacts independently and hence could not perform offline deduplication. Does Thanos receiver not work with HA prometheus remote writers?

**Comment by fpetkovski** [comment]
> Vertical compaction run by Compactor does identifies the blocks from each receiver is different stream

This should not really happen. You should add receiver labels as replica labels in the compactor to make sure blocks end up in the same stream.

Keep in mind that you will still remote-write twice the amount of samples since Receivers cannot deduplicate on write. Have you considered uploading blocks directly to object storage using the sidecar instead?

**Comment by v1jayr** [comment]
Adding all receiver labels as replica label in compactor would result in *all* blocks to get merged into single block, if I understand vertical compaction correctly? We have a number of HA prometheus groups running and thought having 1 block per shard (post vertical compaction) would be better from query performance perspective (assuming it reduces the number of chunks to download as chunks holds only timeseries from a single shard vs all shards). Is this not the case?

**Comment by fpetkovski** [comment]
> Adding all receiver labels as replica label in compactor would result in all overlapping timerange blocks to get merged into single block.

Yes, this is correct. Are you running receivers with a replication factor?

**Comment by 115100** [comment]
We're seeing this issue too. From what I can see, compaction label removing only works on the external labels set in the meta.json [1].

This is only set by the receiver; the external label set in the remote write isn't present in meta.json. Is there some way to either drop Prometheus-level labels during compaction, or set the external label from Prometheus during receiving the remote write?

[1] https://github.com/thanos-io/thanos/blob/2ab0ec27f129d27b9b341ce25ff26f7763b195c6/pkg/block/fetcher.go#L707
